Mathlib/GroupTheory/SpecificGroups/Coxeter.lean

```diff
@@ -0,0 +1,461 @@
+def Matrix.CoxeterGroup := PresentedGroup <| CoxeterGroup.Relations.toSet M
+
+instance : Group (Matrix.CoxeterGroup M) := by
+  exact QuotientGroup.Quotient.group _
+
+namespace CoxeterGroup
+
+def of (b : B) : Matrix.CoxeterGroup M := PresentedGroup.of b
+
+@[simp]
+theorem of_apply (b : B) : of M b = PresentedGroup.of (rels := Relations.toSet M) b :=
+  rfl
+
+end CoxeterGroup
+
+/-- A Coxeter system `CoxeterSystem W` is a structure recording the isomorphism between
+a group `W` and the group presentation corresponding to a Coxeter matrix. Equivalently, this can
+be seen as a list of generators of `W` parameterized by the underlying type of `M`, which
+satisfy the relations of the Coxeter matrix `M`. -/
+structure CoxeterSystem (W : Type*) [Group W]  where
+  /-- `CoxeterSystem.ofMulEquiv` constructs a Coxeter system given an equivalence with the group
+  presentation corresponding to a Coxeter matrix `M`. -/
+  ofMulEquiv ::
+    /-- `mulEquiv` is the isomorphism between the group `W` and the group presentation
+    corresponding to a Coxeter matrix `M`. -/
+    mulEquiv : W ≃* Matrix.CoxeterGroup M
+
+/-- Coxeter system of matrix `M` on the group `M.CoxeterGroup`. -/
+def Matrix.coxeterSystemCoxeterGroup : CoxeterSystem M M.CoxeterGroup :=
+  CoxeterSystem.ofMulEquiv (MulEquiv.refl _)
+
+/-- A group is a Coxeter group if it is registered in a Coxeter system
+ for some Coxeter matrix `M`. -/
+class IsCoxeterGroup (W : Type u) [Group W] : Prop where
+  nonempty_system : ∃ (B : Type u), ∃ (M : Matrix B B ℕ),
+    M.IsCoxeter ∧ Nonempty (CoxeterSystem M W)
+
+namespace CoxeterSystem
+
+open Matrix
+
+variable {B B' W H : Type*} [Group W] [Group H]
+
+variable {M : Matrix B B ℕ} [M.IsCoxeter]
+
+/-- A Coxeter system for `W` with Coxeter matrix `M` indexed by `B`, is associated to
+a map `B → W` recording the images of the indices. -/
+instance funLike : FunLike (CoxeterSystem M W) B (fun _ => W) where
+  coe cs := fun b => cs.mulEquiv.symm (.of b)
+  coe_injective' := by
+    rintro ⟨cs⟩ ⟨cs'⟩ hcs'
+    have H : (cs.symm : CoxeterGroup M →* W) = (cs'.symm : CoxeterGroup M →* W) := by
+      unfold CoxeterGroup
+      ext i; exact congr_fun hcs' i
+    have : cs.symm = cs'.symm := by ext x; exact FunLike.congr_fun H x
+    rw [ofMulEquiv.injEq, ← MulEquiv.symm_symm cs, ← MulEquiv.symm_symm cs', this]
+
+@[simp]
+theorem mulEquiv_apply_coe (cs : CoxeterSystem M W) (b : B) :
+    cs.mulEquiv (cs b) = .of b :=
+  (MulEquiv.eq_symm_apply cs.mulEquiv).mp rfl
+
+/-- The map sending a Coxeter system to its associated map `B → W`. -/
+theorem ext' {c d : CoxeterSystem M W} (H : ⇑c = ⇑d) : c = d := FunLike.coe_injective H
+
+/-- Extensionality rule for Coxeter systems. -/
+theorem ext {c d : CoxeterSystem M W} (H : ∀ x, c x = d x) : c = d :=
+  ext' <| by ext; apply H
+
+/-- The canonical Coxeter system of the Coxeter group over `X`. -/
+def ofCoxeterGroup (X : Type*) (D : Matrix X X ℕ) : CoxeterSystem D (CoxeterGroup D) :=
+  ofMulEquiv (MulEquiv.refl _)
```

Comment:
Isn't this a duplicate of `Matrix.coxeterSystemCoxeterGroup`?

Reply:
Yes it is, I am going to remove the former (while keeping your change from there)